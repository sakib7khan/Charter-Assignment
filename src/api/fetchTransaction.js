export const fetchTransactions = async () => {
  const res = await fetch('http://localhost:3005/transactions');
  if (!res.ok) throw new Error('API fetch failed');
  return await res.json();
};
