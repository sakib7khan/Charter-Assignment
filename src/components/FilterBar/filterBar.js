import React, { useEffect, useState } from 'react';
import { fetchTransactions } from '../../api/fetchTransaction';
import {
  Select,
  Container,
  Spinner,
  ErrorMessage,
} from '../../styles/stylesComponents';

const FilterBar = ({
  selectedMonth,
  selectedYear,
  onMonthChange,
  onYearChange,
}) => {
  const [months, setMonths] = useState([]);
  const [years, setYears] = useState([]);
  const [loading, setLoading] = useState(true);
  const [error, setError] = useState(null);

  useEffect(() => {
    fetchTransactions()
      .then((data) => {
        const monthSet = new Set();
        const yearSet = new Set();

        data.forEach((txn) => {
          const date = new Date(txn.date);
          const monthName = date.toLocaleString('default', { month: 'long' });
          const year = date.getFullYear();
          monthSet.add(monthName);
          yearSet.add(year);
        });

        // Sort months based on calendar order
        const allMonths = [
          'January',
          'February',
          'March',
          'April',
          'May',
          'June',
          'July',
          'August',
          'September',
          'October',
          'November',
          'December',
        ];

        const sortedMonths = allMonths.filter((m) => monthSet.has(m));
        const sortedYears = Array.from(yearSet).sort((a, b) => b - a);

        setMonths(sortedMonths);
        setYears(sortedYears);
        setLoading(false);
      })
      .catch(() => {
        setError('Failed to load filters.');
        setLoading(false);
      });
  }, []);

  if (loading) return <Spinner>Loading filters...</Spinner>;
  if (error) return <ErrorMessage>{error}</ErrorMessage>;

  return (
    <Container>
      <Select
        data-testid='month-select'
        value={selectedMonth}
        onChange={(e) => onMonthChange(e.target.value)}
      >
        <option value=''>All Months</option>
        {months.map((month) => (
          <option key={month} value={month}>
            {month}
          </option>
        ))}
      </Select>

      <Select
        data-testid='year-select'
        value={selectedYear}
        onChange={(e) => onYearChange(e.target.value)}
      >
        <option value=''>All Years</option>
        {years.map((year) => (
          <option key={year} value={year}>
            {year}
          </option>
        ))}
      </Select>
    </Container>
  );
};

export default FilterBar;
