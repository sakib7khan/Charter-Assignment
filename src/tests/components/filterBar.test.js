import React from 'react';
import { render, screen, waitFor, fireEvent } from '@testing-library/react';
import FilterBar from '../../components/FilterBar/filterBar';
import { fetchTransactions } from '../../api/fetchTransaction';

// Mock the fetchTransactions API
jest.mock('../../api/fetchTransaction');

const mockTransactions = [
  { customerId: '1', amount: 120, date: '2024-01-15' },
  { customerId: '1', amount: 90, date: '2025-02-20' },
  { customerId: '2', amount: 150, date: '2025-03-10' },
];

const setup = (props = {}) => {
  return render(
    <FilterBar
      selectedMonth={props.selectedMonth || ''}
      selectedYear={props.selectedYear || ''}
      onMonthChange={props.onMonthChange || jest.fn()}
      onYearChange={props.onYearChange || jest.fn()}
    />
  );
};

describe('FilterBar Component', () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

 
  it('renders loading state initially', async () => {
    fetchTransactions.mockReturnValue(new Promise(() => {})); // never resolves
    setup();
    expect(screen.getByText(/Loading filters/i)).toBeInTheDocument();
  });

  it('renders error state if fetch fails', async () => {
    fetchTransactions.mockRejectedValue(new Error('Failed to fetch'));
    setup();

    await waitFor(() => {
      expect(screen.getByText(/Failed to load filters/i)).toBeInTheDocument();
    });
  });

  it('renders month and year options after successful fetch', async () => {
    fetchTransactions.mockResolvedValue(mockTransactions);
    setup();

    await waitFor(() => {
      expect(screen.getByTestId('month-select')).toBeInTheDocument();
      expect(screen.getByTestId('year-select')).toBeInTheDocument();
    });

    const monthOptions = screen.getAllByRole('option');
    expect(monthOptions.length).toBeGreaterThan(2); // All + at least 2 options

    expect(screen.getByText('January')).toBeInTheDocument();
    expect(screen.getByText('February')).toBeInTheDocument();
    expect(screen.getByText('March')).toBeInTheDocument();
    expect(screen.getByText('2024')).toBeInTheDocument();
    expect(screen.getByText('2025')).toBeInTheDocument();
  });

  it('calls onMonthChange when month is changed', async () => {
    const onMonthChangeMock = jest.fn();
    fetchTransactions.mockResolvedValue(mockTransactions);
    setup({ onMonthChange: onMonthChangeMock });

    await waitFor(() => screen.getByTestId('month-select'));

    fireEvent.change(screen.getByTestId('month-select'), {
      target: { value: 'January' },
    });

    expect(onMonthChangeMock).toHaveBeenCalledWith('January');
  });

  it('calls onYearChange when year is changed', async () => {
    const onYearChangeMock = jest.fn();
    fetchTransactions.mockResolvedValue(mockTransactions);
    setup({ onYearChange: onYearChangeMock });

    await waitFor(() => screen.getByTestId('year-select'));

    fireEvent.change(screen.getByTestId('year-select'), {
      target: { value: '2025' },
    });

    expect(onYearChangeMock).toHaveBeenCalledWith('2025');
  });
});
