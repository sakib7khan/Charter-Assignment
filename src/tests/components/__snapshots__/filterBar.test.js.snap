// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FilterBar Component matches the snapshot 1`] = `
<DocumentFragment>
  <fil />
</DocumentFragment>
`;
